File: cdr/Server/CdrCmd.rb
#!/usr/local/bin/ruby
#----------------------------------------------------------------------
#
# $Id: CdrCmd.rb,v 1.2 2008-02-24 23:04:40 bkline Exp $
#
# Test client (Ruby version) for sending commands to CDR server.
#
# Usage:
#  ruby CdrCmd.rb [command-file [host [port]]]
#
# Example:
#  ruby CdrCmd.rb CdrCommandSamples.xml mmdb2
#
# Default for host is "localhost"; default for port is 2019.
# If no command-line arguments are given, commands are read from standard
# input.  Command buffer must be valid XML, conforming to the DTD
# CdrCommandSet.dtd, and the top-level element must be <CdrCommandSet>.
# The encoding for the XML must be UTF-8.
#
#----------------------------------------------------------------------

#----------------------------------------------------------------------
# Import required packages.
#----------------------------------------------------------------------
require 'socket'

#----------------------------------------------------------------------
# Establish parameters.
#----------------------------------------------------------------------
name = if ARGV.size > 0 then ARGV.shift else '' end
host = if ARGV.size > 0 then ARGV.shift else 'mahler.nci.nih.gov' end
port = if ARGV.size > 0 then ARGV.shift else 2019 end


#----------------------------------------------------------------------
# Get the commands.
#----------------------------------------------------------------------
file = if name == '' then $stdin else open(name) end
cmds = file.read()

#----------------------------------------------------------------------
# Submit them to the CDR server.
#----------------------------------------------------------------------
sock = TCPSocket.open(host, port)
sock.send([cmds.length()].pack("N"), 0)
sock.send(cmds, 0)

#----------------------------------------------------------------------
# Read the server's response.
#----------------------------------------------------------------------
resp = sock.recv(sock.recv(4, 0).unpack("N")[0], 0).chomp()
$stdout << "<!-- *** SERVER RESPONSE *** -->\n" << resp << "\n"

File: Server/CdrCmd.pl
#----------------------------------------------------------------------
#
# $Id: CdrCmd.pl,v 1.1 2001-11-28 19:31:03 bkline Exp $
#
# Test client (Perl version) for sending commands to CDR server.
#
# Usage:
#  perl CdrCmd.pl [command-file [host [port]]]
#
# Example:
#  perl CdrCmd.pl CdrCommandSamples.xml mmdb2
#
# Default for host is "localhost"; default for port is 2019.
# If no command-line arguments are given, commands are read from standard
# input.  Command buffer must be valid XML, conforming to the DTD
# CdrCommandSet.dtd, and the top-level element must be <CdrCommandSet>.
# The encoding for the XML must be UTF-8.
#
#----------------------------------------------------------------------

#----------------------------------------------------------------------
# Import required packages.
#----------------------------------------------------------------------
use IO::Socket;

#----------------------------------------------------------------------
# Establish parameters.
#----------------------------------------------------------------------
my $name = $#ARGV >= 0 ? $ARGV[0] : '-';
my $host = $#ARGV >= 1 ? $ARGV[1] : 'localhost';
my $port = $#ARGV >= 2 ? $ARGV[2] : 2019;

#----------------------------------------------------------------------
# Get the commands.
#----------------------------------------------------------------------
open(CMDS, $name) or die "$name: $!";
my $cmds = '';
my $part;
while (read(CMDS, $part, 2048) > 0) { $cmds .= $part; }

#----------------------------------------------------------------------
# Submit them to the CDR server.
#----------------------------------------------------------------------
my $sock = IO::Socket::INET->new(PeerAddr => $host,
                                 PeerPort => $port,
                                 Type     => SOCK_STREAM) or die $@;
$sock->send(pack("N", length($cmds)), 0);
$sock->send($cmds, 0);

#----------------------------------------------------------------------
# Read the server's response.
#----------------------------------------------------------------------
my $resp = '    ';
$sock->recv($resp, 4, 0) or die "can't read response length: $!";
$sock->recv($resp, unpack("N", $resp), 0) or die "can't read response: $!";
chomp $resp;
print "<!-- *** SERVER RESPONSE *** -->\n$resp\n";
